src/BinaryPack/Extensions/System/TypeExtensions.cs

```diff
@@ -56,17 +55,65 @@ orderby member.Name
                       (mode.HasFlag(SerializationMode.NonPublicMembers) || member.IsPublic() ||
                        member.IsDefined(typeof(SerializableMemberAttribute))) &&
                       (member is FieldInfo fieldInfo && !fieldInfo.IsInitOnly && !fieldInfo.IsLiteral ||
-                       member is PropertyInfo propertyInfo && propertyInfo.CanRead && propertyInfo.CanWrite
+                       member is PropertyInfo propertyInfo && propertyInfo.CanRead
                        && propertyInfo.GetIndexParameters().Length == 0)
-                orderby member.Name
                 select member
-            }).ToArray();
+            }).SortMemberAsConstructorParameters(type)
+              .ToArray();
 
             return members;
         }
 
+        
+        /// Finding a constructor where all serialized members have a parameter with same type and same name (ignoring case)
+        private static IEnumerable<MemberInfo> SortMemberAsConstructorParameters(this IEnumerable<MemberInfo> members, Type type)
+        {
+            // Checking for constructor with no parameter and even any constructors first
+            if (type.GetConstructor(Type.EmptyTypes) != null || type.GetConstructors().Length == 0)
+            {
+                return members.OrderBy(member => member.Name);
```

Comment:
If a parameterless constructor will be used for deserialization, the read-only properties cannot be serialized and should be filtered out here.